code/controllers/subsystem/air.dm

```diff
@@ -77,9 +77,9 @@ SUBSYSTEM_DEF(air)
 	var/timer = TICK_USAGE_REAL
 	var/delta_time = wait * 0.1
 
-	if(currentpart == SSAIR_REBUILD_PIPENETS)
-		if(!resumed) //We track the use for each subprocess for the full tick, so we can get an acccurate idea of how much each part costs
-			cached_cost = 0
+	// Every time we fire, we want to make sure pipenets are rebuilt. The game state could have changed between each fire() proc call
+	// and anything missing a pipenet can lead to unintended behaviour at worse and various runtimes at best.
+	if(length(pipenets_needing_rebuilt))
```

Comment:
I see the issue you mean, if the rebuild queue finishes, then AT takes 100ms to finish, AM will have a good chance of having broken pipenets. Same if slightly less likely for pipenet processing and such.

This fix concerns me a bit though, because it could lead to say, a singularity causing SSatmos to stall forever, constantly doing pipenet rebuilds. In dire need of some testing. If it ends up being the case, this runtime isn't worth fixing, but I hope it doesn't come to that.

I need to read over HE code again, in a perfect world the HE pipe wouldn't have to behave this way, but \<Insert Potion Seller Copypasta Here\>.
Note, I say this under the assumption that the offending runtime source is HE pipes, if it's HE components, I might have some ideas.

Reply:
It's HE pipes for one. Gas flow meters for a second. Both of them call some form of return_air() (HE pipes call their own, meters call their return_air of the thing they're attached to) during their process_atmos() that will runtime on things with null pipelines.